### ISSUE TITLE
[Bug] Restoring Optimizer state causes NaN loss values after resuming Gemma 4 SFT from checkpoint

### ISSUE THREAD

**Issue by vikalluru** [NVIDIA-NeMo/RL#3544]
## Describe the bug

Resuming Gemma 4 SFT from a checkpoint saved with optimizer state causes the loss to become `NaN` during the first one or two optimizer steps after restart.

The checkpointed model weights are finite, and the first forward pass immediately after loading is also finite. The failure occurs only after the restored optimizer performs its first update.

Setting `checkpointing.save_optimizer: false` avoids the problem. The same checkpoint then resumes successfully with fresh optimizer and learning-rate scheduler state.

## Steps/Code to reproduce the bug

Environment and training configuration:

- NeMo RL: `v0.7.0`, commit `37cb7de1a`
- Model: `google/gemma-4-31b-it`
- Training: AutoModel SFT
- Parallelism: FSDP2 with DTensor v2
- Checkpointing: PyTorch Distributed Checkpoint
- Precision: BF16 model weights
- Hardware: 32 GB200 GPUs
- Global batch size: 64
- Maximum total sequence length: 32,768

The issue reproduces with both of these mesh configurations:

- TP=1, CP=4, DP=8
- TP=4, CP=1, DP=8

It also reproduces with both Transformer Engine attention and Hugging Face SDPA attention.

Configuration:

```yaml
optimizer:
  name: fused_adam
  lr: 1.0e-6
  master_weights: true
  store_param_remainders: true
  exp_avg_dtype: torch.bfloat16
  exp_avg_sq_dtype: torch.bfloat16
```
```yaml
checkpointing:
  save_optimizer: true
```

Reproduction procedure:

1. Start SFT from the pretrained Gemma 4 checkpoint.
2. Allow training to save a checkpoint at step N.
3. Stop or terminate the job.
4. Resume training from the step-N checkpoint with optimizer restoration enabled.
5. Observe the loss during the first two optimizer steps after resume.

Example:

```text
step 1501: loss = 1.6389
step 1502: loss = NaN
```

This reproduced across five independent training datasets. Every fresh run completed normally, while every run restored with `save_optimizer: true` failed within zero to two steps.

## Expected behavior

Restoring the checkpoint should reproduce the saved model, optimizer, scheduler, and training state. Training should continue from step N without introducing non-finite parameters or loss values.

The following checks pass immediately after loading:

- All restored model parameters are finite.
- A scan of 4,565 DCP shards covering 31,273,088,876 parameters found no `NaN`, `Inf`, or all-zero tensors.
- Consolidated Hugging Face model weights are also finite.
- The first forward pass after loading produces a finite loss.
- Vision and audio towers are explicitly frozen.

The loss becomes `NaN` only after:

1. Backward
2. Gradient clipping with `clip_grad_norm_`
3. The first restored `optimizer.step()`
4. The next forward pass

This localizes the failure to the first parameter update using the restored optimizer state. Workaround config below:

```yaml
checkpointing:
  save_optimizer: false
  save_consolidated: true
```

## Additional context

Working hypothesis: Problem reconstructing or mapping Transformer Engine FusedAdam state under FSDP2 and PyTorch Distributed Checkpointing. Possible areas include master weights, parameter remainders, BF16 optimizer moments, or optimizer-state-to-parameter mapping.

**Comment by sharonyu-115** [comment]
Hi @vikalluru, I have reviewed the issue and got the following questions so as to have a better understanding of the error scenario: 

1. Which entry point are you using?
`examples/run_sft.py` (NeMo-RL), or `nemo_automodel/recipes/llm/train_ft.py` (Automodel's own SFT recipe)?
I ask because some of the configurations mentioned in the issue are not supported today in the NeMo RL sft path, e.g. CP>1, TP>1. 

2. Would you share the full config YAML used for the runs? 

Thank you.

**Comment by stevenmsm** [comment]
> Hi [@vikalluru](https://github.com/vikalluru), I have reviewed the issue and got the following questions so as to have a better understanding of the error scenario:
> 
> 1. Which entry point are you using?
>    `examples/run_sft.py` (NeMo-RL), or `nemo_automodel/recipes/llm/train_ft.py` (Automodel's own SFT recipe)?
>    I ask because some of the configurations mentioned in the issue are not supported today in the NeMo RL sft path, e.g. CP>1, TP>1.
> 2. Would you share the full config YAML used for the runs?
> 
> Thank you.

Hi @sharonyu-115 and @vikalluru , please see my responses below:

Hi — thanks for the follow-up.

Question 1:
Entry point: examples/run_sft.py (NeMo-RL). Our launch wrapper invokes it directly:
```
uv run --no-sync ${NEMO_RL_HOME}/examples/run_sft.py \
    --config <sft_yaml> \
    <cluster.* / policy.* overrides>
```
where NEMO_RL_HOME=/opt/nemo-rl inside our container (nemo-rl commit 37cb7de1a). We are not using nemo_automodel/recipes/llm/train_ft.py.

On the CP>1 / TP>1 caveat: We opt into DTensor v2 via the yaml, which routes to DTensorPolicyWorkerV2 and the automodel-integrated backend. The relevant policy block:
```
policy:
  dtensor_cfg:
    _v2: true
    enabled: true
    tensor_parallel_size: 1
    context_parallel_size: 4
    sequence_parallel: false
    cpu_offload: true
    activation_checkpointing: true
```
Empirically this configuration trains cleanly from scratch — we've completed many independent fresh SFT runs on Gemma-4-31B E2E each with this exact TP=1, CP=4, DP=8 mesh, no NaN, healthy loss curves. The NaN reproduces only on resume from a checkpoint saved with save_optimizer: true.
The TP=4, CP=1 mesh mentioned in the report was also tested, and the same resume-NaN issue occurs. 

Thus, the loss NaN when loading checkpoint with save_optimizer=True is reproducible on both meshes, i.e. it is mesh-independent. The issue goes away when we set save_optimizer: false on our SFT config, and this is our current workaround.

Question 2:
```
sft:
  ## total number of steps to train will equal
  ## min((max_num_epochs * len(train_dataloader)), max_num_steps)
  max_num_epochs: 1
  max_num_steps: 999999999

  val_period: 200
  val_batches: 8
  val_global_batch_size: 64
  val_micro_batch_size: 1
  val_at_start: true
  val_at_end: false
  seed: 42

checkpointing:
  enabled: true
  checkpoint_dir: "results/sft-gemma-4-31b"
  metric_name: null ## set to null to save most recent k checkpoints
  higher_is_better: false
  keep_top_k: 2
  save_period: 200
  checkpoint_must_save_by: null
  save_optimizer: false # if true will cause loss NaN issue at step N+1
  # save_consolidated MUST be false at training time for 31B on GB200: the
  # rank-0 all-gather to write HF-format safetensors hangs at dist.barrier() in
  # nemo_automodel/components/checkpoint/_backports/consolidate_hf_safetensors.py:855
  # and overruns the 30-min Gloo TCP timeout. Confirmed at both SFT (step 200
  # save, sft_mix-y2.5-1p-v1) and DPO. Same fix pattern as colleague's GRPO
  # Gemma-4 config. To get HF-format weights for eval, run
  # train_dpo/oci/gemma_4_31b/consolidate_only.py post-training on the DCP
  # shards — dedicated job, no training-loop watchdog constraint.
  save_consolidated: false

policy:
  model_name: "<path to gemma-4-31B>" # e.g. /mnt_gb200/users/simin.ma/data/genAI/public_models/gemma-4-31B
  tokenizer:
    name: ${policy.model_name} ## use the base model's tokenizer
    # Gemma-4-31B base likely ships without chat_template.jinja. Options:
    #   1) copy the jinja from google/gemma-4-31B-it into the base model dir
    #      and set chat_template: ${policy.model_name}/chat_template.jinja
    #   2) point directly at an -it checkpoint's chat_template.jinja
    #   3) inline a jinja string here
    chat_template: ${policy.model_name}/chat_template.jinja
    chat_template_kwargs: null

  train_global_batch_size: 64
  train_micro_batch_size: 1
  # Gemma-4 supports 262144 context; 32K is our SFT target. Override via CLI
  # if you need more/less.
  max_total_sequence_length: 32768
  # Chunk logprob computation to bound peak activation memory on long sequences
  # (matches the -it DAPO recipe).
  logprob_chunk_size: 4096
  precision: "bfloat16"

  offload_optimizer_for_logprob: true

  dtensor_cfg:
    _v2: true
    enabled: true
    env_vars:
      PYTORCH_CUDA_ALLOC_CONF: "expandable_segments:True"
    cpu_offload: true
    sequence_parallel: false
    activation_checkpointing: true
    tensor_parallel_size: 1
    context_parallel_size: 4
    custom_parallel_plan: null

    # nemo-automodel loader kwargs. freeze_config is REQUIRED for Gemma-4 to
    # keep the SFT loss/optimizer from touching the vision + audio towers.
    automodel_kwargs:
      attn_implementation: "te"
      use_liger_kernel: false
      use_sdpa_patching: false
      freeze_config:
        freeze_vision_tower: true
        freeze_audio_tower: true
        freeze_language_model: false

  dynamic_batching:
    enabled: true
    train_mb_tokens: ${mul:${policy.max_total_sequence_length}, ${policy.train_micro_batch_size}}
    sequence_length_round: 64

  # DTensor path — seq len must be divisible by (2 * TP * CP). The 2x factor
  # comes from PyTorch's CP zigzag load balancer at
  # torch/distributed/tensor/experimental/_context_parallel/_load_balancer.py:148:
  #   assert seq_length % (world_size * 2) == 0
  # which splits the seq into 2*cp_size chunks to balance the causal-mask work.
  # For TP=1, CP=2 -> divisor 4; for TP=4, CP=2 -> divisor 16. 32K covers both.
  make_sequence_length_divisible_by: ${mul:2, ${mul:${policy.dtensor_cfg.tensor_parallel_size}, ${policy.dtensor_cfg.context_parallel_size}}}
  max_grad_norm: 1.0

  # TE FusedAdam with bf16 momentum + master weights — matches the official
  # gemma-4-31b-it DAPO recipe and saves optimizer state memory.
  #
  # LR default is set for "light SFT from -it": 1e-6 sits below Llama-70B's 2e-6
  # (from-base) because we're gently nudging an already-instruct model and
  # actively trying not to forget. train_sft_base.sh overrides this per run.
  optimizer:
    name: transformer_engine.pytorch.optimizers.fused_adam.FusedAdam
    kwargs:
      lr: 1.0e-6
      weight_decay: 0.1
      betas: [0.9, 0.98]
      eps: 1.0e-8
      master_weights: true
      store_param_remainders: true
      exp_avg_dtype: torch.bfloat16
      exp_avg_sq_dtype: torch.bfloat16

  # Scheduler: LinearLR warmup -> CosineAnnealingLR decay, chained via SequentialLR
  # (torch's list-of-schedulers with milestones is what nemo-rl automodel expects).
  #   [0] LinearLR: LR climbs from 1% to 100% of the base LR over `total_iters` steps.
  #   [1] CosineAnnealingLR: decays over `T_max` steps to `eta_min`.
  #   [2] milestones = [warmup_iters]  -> switches from LinearLR to CosineAnnealingLR.
  # train_sft_base.sh computes warmup_iters (~4% of total_iters, floor 5) and
  # T_max (= total_iters - warmup_iters) and overrides these fields at runtime.
  # The defaults below are placeholders for a "no data yet" launch.
  scheduler:
  - name: torch.optim.lr_scheduler.LinearLR
    kwargs:
      start_factor: 1.0e-2
      end_factor: 1.0
      total_iters: 20
  - name: torch.optim.lr_scheduler.CosineAnnealingLR
    kwargs:
      T_max: 500
      eta_min: 1.0e-7
  - milestones:
    - 20

## Refer to https://docs.nvidia.com/nemo/rl/latest/guides/sft.html for new format of data config
## Also: https://github.com/NVIDIA-NeMo/RL/pull/1649
data:
  max_input_seq_length: ${policy.max_total_sequence_length}
  add_bos: true
  add_eos: true
  add_generation_prompt: false
  shuffle: true
  num_workers: 4
  train:
    data_path: "data/train.jsonl"
    split_validation_size: 0.0
  validation:
    data_path: "data/val.jsonl"
  ## default settings for all datasets
  default:
    dataset_name: "openai_format"
    chat_key: "messages"
    processor: "sft_processor"
    system_key: null
    system_prompt: null
    system_prompt_file: null
    tool_key: "tools"
    use_preserving_dataset: false
    prompt_file: null
    # Empty strings satisfy the Pydantic BaseModel validation added in
    # upstream PR #2325 (ResponseDatasetConfig / PreferenceDatasetConfig now
    # enforce string type) while preserving the "unused" intent for JSONL
    # data — see sft_nemotron3_nano_30b.yaml for context.
    split: ""
    output_key: ""
    seed: 42

logger:
  log_dir: "logs"
  wandb_enabled: true
  tensorboard_enabled: true
  mlflow_enabled: false
  monitor_gpus: true
  swanlab_enabled: false
  num_val_samples_to_print: 0
  wandb:
    project: "sft-gemma-4-31b"
    name: ${logger.wandb.project}
  tensorboard:
    log_dir: tb_logs-${logger.wandb.project}
  mlflow:
    experiment_name: ${logger.wandb.project}
    run_name: ${logger.wandb.name}
  gpu_monitoring:
    collection_interval: 10
    flush_interval: 10

cluster:
  gpus_per_node: 8
  num_nodes: 4
```

**Comment by sharonyu-115** [comment]
Hi @stevenmsm, Thank you for the detailed information. It seems like the problem is related to FusedAdam + cpu_offload: true. I haven't confirmed the root cause yet, but wanted to share the findings here in case you have time to test it out by either disabling cpu_offload or change to use Adam.

**Comment by sharonyu-115** [comment]
Hi @stevenmsm  @vikalluru , I was able to reproduce the loss nan issue and I think it can conclude the problem is related to FusedAdam used with cpu_offload true. 

<img width="789" height="330" alt="Image" src="https://github.com/user-attachments/assets/f2fcd97e-0e76-4d42-9691-822b31624e34" />

Fused Adam is GPU only https://github.com/NVIDIA/TransformerEngine/blob/366798ef8a0a00d8f2c1650d11e7e623d7c33e26/transformer_engine/pytorch/optimizers/fused_adam.py#L38 , when used with `cpu_offload: true` on GB, it does not error out but corrupt the optimizer moments `w` and `v`.  I examined the moments in the saved checkpoints using the following script:

[scan_optimizer_checkpoint.py](https://github.com/user-attachments/files/31066180/scan_optimizer_checkpoint.py)

<img width="592" height="94" alt="Image" src="https://github.com/user-attachments/assets/c00a35ad-c09f-4abe-ace6-542a40f65099" />

Results show when cpu_offload is enabled, `exp_avg` is poisoned. 

I also had similar instrument to dump the value during the training run, after `optimizer.step()`, I think the moments are corrupted even in the memory. Though I cannot fully explain why the problem is only visible after resume, the cause of the error is clear I think. Would you give a try on your end? Thanks!

**Comment by stevenmsm** [comment]
Hi @sharonyu-115, thanks for the results, I will try it out as well. Meanwhile, since we are training SFT Gemma with 32K seq_len, CP=4, TP=1, with 32 GB200 gpus if I do not set cpu_offload as true, it will oom in some batches when the sequence lengths are near 32K. Thus, I might be only able to try FusedAdam -> Adam. Would this work as well?

**Comment by sharonyu-115** [comment]
hi @stevenmsm , yes, Adam works with cpu_offload.

**Comment by stevenmsm** [comment]
@sharonyu-115 I am able to terminate and recontinue training with save_optimizer=True by changing FusedAdam to Adam or AdamW. Thanks for the help.
